link/kprobe_test.go

```diff
@@ -0,0 +1,241 @@
+	c.Assert(err, qt.IsNil)
+	defer pe.Close()
+
+	// kretprobe happy path.
+	pr, err := pmuKprobe("printk", true)
+	c.Assert(err, qt.IsNil)
+	defer pr.Close()
+
+	c.Assert(pr.progType, qt.Equals, ebpf.Kprobe)
+
+	pe, err = pr.perfEventOpenPMU()
+	c.Assert(err, qt.IsNil)
+	defer pe.Close()
+}
+
+func BenchmarkKprobeCreatePMU(b *testing.B) {
+	pr, err := pmuKprobe("printk", false)
+	if err != nil {
+		b.Error("error getting kprobe PMU type:", err)
+	}
+	defer pr.Close()
+
+	for n := 0; n < b.N; n++ {
+		fd, err := pr.perfEventOpenPMU()
+		if err != nil {
+			b.Error("error creating PMU perf event:", err)
+		}
+
+		if err := fd.Close(); err != nil {
+			b.Error("error closing PMU perf event:", err)
+		}
+	}
+}
+
+// Test tracefs k(ret)probe creation on all kernel versions.
+// Create two trace events on the same symbol and ensure their IDs differ.
+func TestKprobeTraceFS(t *testing.T) {
+	c := qt.New(t)
+
+	symbol := "printk"
+
+	// Open and close tracefs kprobe, checking all errors.
+	te, err := tracefsKprobe(symbol, false)
+	c.Assert(err, qt.IsNil)
+	c.Assert(te.Close(), qt.IsNil)
+
+	// Create similar trace events, ensure their IDs differ.
+	te1, err := tracefsKprobe(symbol, false)
+	c.Assert(err, qt.IsNil)
+	defer te1.Close()
+	c.Assert(te1.progType, qt.Equals, ebpf.Kprobe)
+	c.Assert(te1.tracefsID, qt.Not(qt.Equals), 0)
+
+	te2, err := tracefsKprobe(symbol, false)
+	c.Assert(err, qt.IsNil)
+	defer te2.Close()
+	c.Assert(te2.progType, qt.Equals, ebpf.Kprobe)
+	c.Assert(te2.tracefsID, qt.Not(qt.Equals), 0)
+
+	c.Assert(te1.tracefsID, qt.Not(qt.CmpEquals()), te2.tracefsID)
+
+	// Write a kprobe event for a non-existing symbol.
+	err = createTraceFSKprobeEvent("syscalls", "bogus", false)
+	c.Assert(errors.Is(err, ErrNotSupported), qt.IsTrue)
+}
+
+func BenchmarkKprobeCreateTraceFS(b *testing.B) {
+	for n := 0; n < b.N; n++ {
+
+		// Include <tracefs>/kprobe_events operations in the benchmark loop
+		// because we create one per perf event.
+		pr, err := tracefsKprobe("printk", false)
+		if err != nil {
+			b.Error("error creating tracefs trace event:", err)
+		}
+		defer pr.Close()
+
+		fd, err := pr.perfEventOpenTraceFS()
+		if err != nil {
+			b.Error("error creating tracefs-backed perf event:", err)
+		}
+
+		if err := fd.Close(); err != nil {
+			b.Error("error closing tracefs perf event:", err)
+		}
+	}
+}
+
+// Test k(ret)probe creation writing durectly to <tracefs>/kprobe_events.
+// Only runs on 5.0 and over. Earlier versions ignored writes of duplicate
+// events, while 5.0 started returning -EEXIST when a kprobe event already
+// exists.
+func TestKprobeCreateTraceFS(t *testing.T) {
+	testutils.SkipOnOldKernel(t, "5.0", "<tracefs>/kprobe_events doesn't reject duplicate events")
+
+	c := qt.New(t)
+
+	symbol := "printk"
+	pg := mustRandomGroup("ebpftest")
+	rg := mustRandomGroup("ebpftest")
+
+	// Tee up cleanups in case any of the Asserts abort the function.
+	defer func() { _ = closeTraceFSKprobeEvent(pg, symbol) }()
```

Comment:
Just deferring them directly doesn't work?

Reply:
It does, but then we'd need to disable errcheck for those lines. This is fine.

Reply:
Ah, the false positive strikes again.